File: conformance/code/individual-go/internal/clients/idgen_client.go
package clients

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"

	"individual/internal/config"

	"github.com/google/uuid"
	"github.com/rs/zerolog/log"
)

// IDGenClient handles ID generation service integration
type IDGenClient interface {
	// Added optional customVars to allow passing ORG, etc.
	GenerateIDs(ctx context.Context, tenantID string, idFormat string, count int, customVars map[string]string) ([]string, error)
}

type idGenClient struct {
	config     *config.IDGenConfig
	httpClient *http.Client
	enabled    bool
}

// NewIDGenClient creates a new ID generation client
func NewIDGenClient(cfg *config.IDGenConfig) IDGenClient {
	return &idGenClient{
		config:     cfg,
		httpClient: &http.Client{},
		enabled:    cfg.Enabled,
	}
}

// localIDGenRequest matches the local IdGen service contract
type localIDGenRequest struct {
	TemplateCode string            `json:"templateCode"`
	Variables    map[string]string `json:"variables"`
}

// localIDGenResponse matches the local IdGen service response
type localIDGenResponse struct {
	ID string `json:"id"`
}

// GenerateIDs generates formatted IDs from IDGen service
// GenerateIDs generates formatted IDs from IDGen service
func (c *idGenClient) GenerateIDs(ctx context.Context, tenantID string, idFormat string, count int, customVars map[string]string) ([]string, error) {
	if !c.enabled {
		log.Warn().Int("count", count).Msg("IDGen disabled, generating fallback IDs")
		return c.generateFallbackIDs(count), nil
	}

	url := c.config.Host + c.config.Path

	// Ensure variables map is initialized
	if customVars == nil {
		customVars = make(map[string]string)
	}

	// Always include ORG (default PG) unless provided
	if _, exists := customVars["ORG"]; !exists {
		customVars["ORG"] = tenantID
	}

	ids := make([]string, 0, count)
	for i := 0; i < count; i++ {
		payload := localIDGenRequest{
			TemplateCode: idFormat,
			Variables:    customVars,
		}

		jsonData, err := json.Marshal(payload)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal IDGen request: %w", err)
		}

		req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewBuffer(jsonData))
		if err != nil {
			return nil, fmt.Errorf("failed to create request: %w", err)
		}
		req.Header.Set("Content-Type", "application/json")
		// Set tenant header as required by IDGen
		if tenantID != "" {
			req.Header.Set("X-Tenant-Id", tenantID)
		}

		// idgen is ENABLED here (disabled short-circuits to fallback above). A failure is returned as
		// an error — we do NOT silently fall back, so a broken dependency fails loudly.
		resp, err := c.httpClient.Do(req)
		if err != nil {
			return nil, fmt.Errorf("failed to call IDGen service: %w", err)
		}

		err = func() error {
			defer resp.Body.Close()

			if resp.StatusCode != http.StatusOK {
				body, _ := io.ReadAll(resp.Body)
				return fmt.Errorf("idgen returned status=%d body=%s", resp.StatusCode, string(body))
			}

			var response localIDGenResponse
			if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
				return fmt.Errorf("failed to decode IDGen response: %w", err)
			}

			if response.ID == "" {
				return fmt.Errorf("idgen returned empty id")
			}

			ids = append(ids, response.ID)
			return nil
		}()
		if err != nil {
			return nil, err
		}
	}

	return ids, nil
}

// generateFallbackIDs generates simple UUID-based IDs when IDGen service is unavailable
func (c *idGenClient) generateFallbackIDs(count int) []string {
	ids := make([]string, count)
	for i := 0; i < count; i++ {
		// Generate unique ID using UUID to avoid duplicates
		ids[i] = "IND-" + uuid.New().String()[:8]
	}
	return ids
}

File: conformance/code/employee/src/main/java/com/digit/employee/client/IdGenClient.java
package com.digit.employee.client;

import com.digit.employee.config.EmployeeProperties;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Client for the IDGen service. Mirrors Go internal/clients/idgen/client.go: POSTs {@code count}
 * times to {@code host + path} with {templateCode, variables{ORG:tenantId}} and collects {@code id}.
 */
@Component
public class IdGenClient {

    private static final Logger log = LoggerFactory.getLogger(IdGenClient.class);

    private final HttpClient httpClient = HttpClient.newHttpClient();
    private final ObjectMapper objectMapper;
    private final String host;
    private final String path;
    private final String idgenName;
    private final boolean enabled;

    public IdGenClient(EmployeeProperties props, ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
        // Trim a trailing slash on host so host + path (e.g. "/idgen/v3/generate") yields a
        // single-slash URL, matching Go idgen client (strings.TrimSuffix(host, "/")).
        String h = props.getIdgen().getHost();
        this.host = (h != null && h.endsWith("/")) ? h.substring(0, h.length() - 1) : h;
        this.path = props.getIdgen().getPath();
        this.idgenName = props.getIdgen().getIdgenName();
        this.enabled = props.getIdgen().isEnabled();
    }

    /**
     * Generates {@code count} formatted IDs. Dependency-flag semantics: when idgen is DISABLED we
     * generate local fallback ids (no error — the service runs standalone). When ENABLED, a failure
     * (network / non-200 / missing id) is surfaced as an error — we do NOT silently fall back, so a
     * broken dependency fails loudly rather than minting non-standard ids.
     */
    public List<String> generateIDs(String tenantId, int count, Map<String, String> customVars) {
        if (!enabled) {
            log.info("IDGen disabled — generating {} local fallback id(s)", count);
            return generateFallbackIds(count);
        }
        String url = host + path;

        // Variables are passed through as-is (Go removed the ORG default injection in 8749c30e).
        Map<String, String> vars = customVars == null ? new HashMap<>() : new HashMap<>(customVars);

        List<String> ids = new ArrayList<>(count);
        for (int i = 0; i < count; i++) {
            Map<String, Object> payload = new HashMap<>();
            payload.put("templateCode", idgenName);
            payload.put("variables", vars);

            try {
                byte[] jsonData = objectMapper.writeValueAsBytes(payload);
                HttpRequest.Builder b = HttpRequest.newBuilder()
                        .uri(URI.create(url))
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofByteArray(jsonData));
                if (tenantId != null && !tenantId.isEmpty()) {
                    b.header("X-Tenant-Id", tenantId);
                }
                HttpResponse<String> resp = httpClient.send(b.build(), HttpResponse.BodyHandlers.ofString());

                if (resp.statusCode() != 200) {
                    // Fail immediately with status + body (matches Go idgen client).
                    throw new RuntimeException("idgen returned status=" + resp.statusCode()
                            + " body=" + resp.body());
                }
                JsonNode node = objectMapper.readTree(resp.body());
                String id = node.has("id") ? node.get("id").asText("") : "";
                if (id.isEmpty()) {
                    throw new RuntimeException("idgen response missing 'id'");
                }
                ids.add(id);
            } catch (RuntimeException e) {
                throw e;
            } catch (Exception e) {
                throw new RuntimeException("failed to call IDGen service: " + e.getMessage(), e);
            }
        }
        return ids;
    }

    /** Local id generation used only when idgen is disabled. */
    private List<String> generateFallbackIds(int count) {
        List<String> ids = new ArrayList<>(count);
        for (int i = 0; i < count; i++) {
            ids.add("EMP-" + java.util.UUID.randomUUID().toString().substring(0, 8));
        }
        return ids;
    }
}
